packages/nox/src/NOX_Exceptions.H

```diff
@@ -0,0 +1,64 @@
+//@HEADER
+// ************************************************************************
+//
+//            NOX: An Object-Oriented Nonlinear Solver Package
+//                 Copyright (2002) Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Roger Pawlowski (rppawlo@sandia.gov) or
+// Eric Phipps (etphipp@sandia.gov), Sandia National Laboratories.
+//@HEADER
+
+#ifndef   NOX_EXCEPTIONS_H
+#define   NOX_EXCEPTIONS_H
+
+namespace NOX {
+namespace Exceptions {
+
+  /*!
+    \brief A generic exception class denoting a solver failure.
+
+    This class is a member of namespace NOX::Exceptions.
+  */
+  class SolverFailure : public std::logic_error {
+
+  public:
+
+    //! Constructor.
+    SolverFailure(const std::string& what) : std::logic_error(what) {};
```

Comment:
Good thought @bartlettroscoe.  @etphipp, does this change sound good to you?

Reply:
NOX still only optionally supports Thyra - introducing that means you would have to move the code into the src-thyra and protect any use of that exception with ifdefs for thyra being enabled.

Reply:
I don't like the sounds of that.  Should we perhaps just remove the Thyra exception altogether and just go with the NOX one?

Reply:
That is probably the best thing to do.

Reply:
This makes the Thyra::LinearOpWtihSolveBase interface (or the NOX implemenation) lie to a Thyra user.  Could this be made to conditnially use std::logic_error or Teuchos::LinearOpApplyFailed?

Reply:
@jmgate, did you rebase this branch to remove the changes to the Thyra::LinearOpBase interface?  I was going to add conditioinal ifdefed code to use the Thyra exception class as the base class when Thyra was enabled but use the std:: exception class when it was not.  That would have satisfied both requirements:

1. Throw an exception of type Thyra::Excpetions::LinearOpApplyFailed when apply fails
2. Not introduce a required dependence of this code on NOX

Unfortunately, the current branch does not satisfy the Thyra::LinearOpBase interface that says that Thyra::LinearOpBase::apply() can never fail (except for clear failure to satisfy the preconditions).  This type of error does not fall into that category.

Actually, when not deriving from Thyra::Exceptions::LinearOpApplyFailed, should not NOX::SolverFailure derive from std::runtime_error instead of std::logic_error?  See:

* http://forums.codeguru.com/showthread.php?461023-std-runtime_error-vs-std-logic_error

And can add a commit that does these things without objection and will satisfy all three requirements.

Reply:
I went ahead and merged this PR because it gives Charon the functionality it needs and has been waiting for since August.  If you guys want to figure out a different implementation with respect to how NOX and Thyra interact, that's fine with me&mdash;just let me know what'll need to change in Charon eventually, if anything.